### ISSUE TITLE
[Bug] Vision-Exp: `bias_vl` is loaded but never read — image tokens route through the text MoE path in all 43 layers

### ISSUE THREAD

**Issue by alaistair76** [MiaAI-Lab/DeepSeek-v4-Flash-DSpark-2x-DGX-Spark#175]
## Summary

The Vision-Exp checkpoint ships a vision-specific MoE router bias
(`.ffn.gate.bias_vl`) on 46 layers. `hotfix-dsv4-vision-exp.py` creates the
parameter and the weights mapper loads it correctly — but nothing ever reads it.
Every image token is therefore routed by the **text** router bias, and in the
three hash-MoE layers every image token collapses onto a single fixed expert set.

Text-only requests are unaffected. This is image-path only.

## Environment

- Recipe: `de230b4` (also present at `d58c877`)
- Model: `deepseek-ai/DeepSeek-V4-Flash-Vision-Exp`, revision `e46e16bf…`
- Image: `ghcr.io/anemll/dspark-vllm-gx10:0.1.1`
- vLLM `0.25.2.dev0+g752a3a504.d20260714`
- 2× DGX Spark, TP=2 over RoCE, `deepseek_v4_fp8`, KV `nvfp4_ds_mla`
- `max_num_seqs=6`, `max_num_batched_tokens=8192`, `num_speculative_tokens=6`

## What the reference implementation does

`inference/model.py:609-633` branches on an image mask in two distinct ways:

```python
self.bias_vl  = nn.Parameter(...) if vl else None
image_mask    = (input_ids >= self.vocab_size) if self.bias_vl is not None else None

# hash layers: image tokens ABANDON hash routing for learned top-k
indices    = self.tid2eid[torch.where(image_mask, 0, input_ids)]
vl_indices = (scores + self.bias_vl).topk(self.topk, dim=-1)[1]
indices    = torch.where(image_mask.unsqueeze(-1), vl_indices.to(indices.dtype), indices)

# non-hash layers: image tokens get bias_vl instead of bias
scores = scores + torch.where(image_mask.unsqueeze(-1), self.bias_vl, self.bias)
```

## What the recipe does

The only routing call is `fused_topk_bias(...)` in
`vllm/models/deepseek_v4/nvidia/model.py`, which receives
`e_score_correction_bias=self.gate.e_score_correction_bias` (the **text** bias),
`input_tokens=input_ids` and `hash_indices_table=self.gate.tid2eid`. There is no
image mask anywhere in the MoE path.

`apply.py`'s `moe_init` creates `self.gate.e_score_correction_bias_vl` so the
checkpoint tensor has somewhere to land, and `VISION_MAPPER_SUFFIXES` remaps
`.ffn.gate.bias_vl` onto it — so it loads cleanly and is then discarded.

`grep -rn bias_vl` across the whole installed vLLM tree returns **four** hits,
all in `nvidia/dspark.py`, all pure name-remapping.

## Why this is not cosmetic

`bias_vl` is present on 46 layers and selects almost entirely different experts
from the text bias (measured directly from the checkpoint):

| layer | rms(vl) | rms(text) | max abs diff | top-8 expert overlap |
|---|---|---|---|---|
| layers.0 | 1.645 | 71.36 | 74.92 | **0/8** |
| layers.1 | 1.267 | 71.42 | 74.45 | **0/8** |
| layers.3 | 1.052 | 10.13 | 9.28 | 2/8 |
| layers.10 | 1.024 | 9.66 | 9.02 | **0/8** |
| layers.30 | 2.285 | 13.25 | 11.66 | 1/8 |

(`mtp.1` / `mtp.2` have `rms(bias_vl) = 0.000`, so the DSpark draft correctly
does no VL routing — the `dspark.py` skip is right.)

## Measured at runtime

One real image request (149 image tokens, 29 text tokens), all 43 MoE layers
instrumented with a read-only wrapper on `DeepseekV4MoE.forward`, 0 trace errors.

**Hash layers 0–2 — total collapse.** `all_image_tokens_same_experts=True`,
`distinct_expert_sets_among_image_tokens=1`:

| layer | expert set every image token receives | text tokens, same forward |
|---|---|---|
| layers.0 | `[62, 254, 135, 159, 170, 174]` | 29 tokens → **21 distinct** sets |
| layers.1 | `[58, 110, 202, 54, 132, 208]` | 21 distinct sets |
| layers.2 | `[52, 187, 245, 55, 182, 246]` | 21 distinct sets |

Layer 0's set is exactly `tid2eid[129264]` — the placeholder id — as predicted
from the checkpoint. Note `129264 < vocab_size 129280`, so the hash lookup is
*valid* and fails silently.

**Non-hash layers 3–42** — real hidden states, real gate,
`topk(scores + bias)` vs `topk(scores + bias_vl)`:

- mean expert overlap **0.970 of 6** → on average **5.03 of every 6 experts differ**
- `frac_image_tokens_identical_routing = 0.000` in **every one of the 40 layers**
- worst at the top of the stack: layers 40/41/42 overlap 0.221 / 0.101 / 0.094,
  where `rms(text bias)` reaches 38.4 against `rms(vl)` ≈ 2.3
- layer 3, one token: ours `[64,68,183,124,119,166]` vs reference `[68,183,14,232,36,194]`

## Root cause

The reference computes its mask as `input_ids >= vocab_size` because it uses
**out-of-vocab sentinels**. The recipe correctly switched to a real in-vocab
placeholder (129264) at the engine boundary — which is the right call for vLLM —
but that silently removed the signal the MoE depended on.

## Why existing tests pass

`scripts/test-dsv4-vision-exp-hotfix.py` covers geometry, `build_image_block`
types/perm, `as_pil` and the role checks. Nothing exercises MoE routing.
Behavioural canaries also pass: the model answered a 2×2 quadrant colour probe
correctly ("red green blue yellow") **during the very trace that showed it
misrouted in all 43 layers**. Coarse image QA survives bad expert selection.

## Suggested fix

Both routing paths must be covered:

1. `DeepseekV4MoE.forward` mega-MoE branch calls `fused_topk_bias` directly.
2. `DeepseekV4MoE._forward_fused_moe` does **not** route locally — it delegates
   to `self.experts(...)`, and `FusedMoE` was handed `e_score_correction_bias`
   and `hash_indices_table` at **construction**. A patch to `forward` alone
   silently misses this path.

The common funnel is `fused_topk_bias`, whose signature takes a single
`e_score_correction_bias` — so the fix has to split rows rather than pass two
biases. Suggested shape: derive `image_mask = (input_tokens == 129264)` (per-chunk
correct by construction), then call `fused_topk_bias` twice on disjoint row
subsets and scatter back — text rows as today, image rows with
`e_score_correction_bias=bias_vl` **and `hash_indices_table=None`**, so hash
layers fall through to learned top-k as the reference does.

## Verification

We have a read-only tracing module that wraps `DeepseekV4MoE.forward`, is
default-off, self-limiting, and reports per-layer expert overlap against what
`bias_vl` would have selected. A correct fix should show
`frac_image_tokens_identical_routing = 1.000` across the non-hash layers and
`distinct_expert_sets_among_image_tokens > 1` in layers 0–2. Happy to share it.

**Comment by MiaAI-Lab** [comment]
Fixed in #179 (`101e6f8` on `main`).

Image placeholder rows (id 129264) now call `fused_topk_bias` with `bias_vl` and no hash table; text rows keep text bias + `tid2eid`. CUDA graph capture takes the stock text path so boot does not `.item()` token ids.

Live on this pair after recreate:
- Both ranks healthy; overlay has `is_current_stream_capturing` + `_wrap_router_compute_routing`
- Solid red JPEG → HTTP 200, 117 image tokens, reply `Red`
- Earth-in-hands PNG → HTTP 200, 365 image tokens, reply describes hands cradling the Earth
- Pair stayed healthy (no EngineCore kill)

Recreate after pull (`./stop` then `./start`); restart keeps old overlay bytes.

**Comment by jmeadlock** [comment]
Verified on 2x DGX Spark (GB10, TP=2, Anemll 0.1.1) at tip 97e8733 with the Keys Vision-EXP-ablit runtime direction (ABLITERATED=1, lambda=3.5):

- The #165/#167/#172/#175/#181 vision fix chain is what makes the ablit lane usable here. The de230b45-era checkout crashed on encoding drift; a tip clone boots clean and passes real image_url, tool-calling, and long-count gates on the ablit weights.
- Block-K unlock (DSPARK_ENABLE_DSPARK_BLOCK_K=1, MTP_NUM_TOKENS=5) measured +2-5% single-stream decode vs K=6 on this pair, KV pool 2.34M tokens.
- SP indexer: no measurable TTFT win at <=48K prefill in our A/B (single cold samples, prefix-cache caveats) — matches the "live A/B still pending" status; parked here.
- One ops note for other ablit users: prepare --abliterated fails if ~/.cache/huggingface/dspark-ablation is root-owned (from an earlier container-run download) — chown fixes it. A permissions check in prepare might save others the same stumble.

The fail-closed hotfix chain and flight-recorder wiring are noticeably better ops than the August revisions. Thanks for the solid work.
